pkg/api/types.go

```diff
@@ -466,6 +470,30 @@ type GlusterfsVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// RBDVolumeSource represents a Rados Block Device Mount that lasts the lifetime of a pod
+type RBDVolumeSource struct {
+	// Required: EndpointsName is the endpoint name that represents a collection of Ceph monitors
+	EndpointsName string `json:"endpoints"`
+	// Required: RBDImage is the rados image name
+	RBDImage string `json:"image"`
+	// Required: Filesystem type to mount.
+	// Must be a filesystem type supported by the host operating system.
+	// Ex. "ext4", "xfs", "ntfs"
+	// TODO: how do we prevent errors in the filesystem from compromising the machine
+	FSType string `json:"fsType,omitempty"`
+	// Optional: RBDPool is the rados pool name,default is rbd
+	RBDPool string `json:"pool"`
+	// Optional: RadosUser is the rados user name, default is admin
+	RadosUser string `json:"user"`
+	// Optional: Keyring is the path to key ring for RBDUser, default is /etc/ceph/keyring
+	Keyring string `json:"keyring"`
+	// Optional: Secret is the authentication secret for RBDUser, default is empty.
+	Secret string `json:"secret" description:"secret is authentication secret. If provided, it overrides keyring"`
```

Comment:
> On Apr 15, 2015, at 3:37 PM, Mark Turansky <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> > +	EndpointsName string `json:"endpoints"`
> > +	// Required: RBDImage is the rados image name
> > +	RBDImage string `json:"image"`
> > +	// Required: Filesystem type to mount.
> > +	// Must be a filesystem type supported by the host operating system.
> > +	// Ex. "ext4", "xfs", "ntfs"
> > +	// TODO: how do we prevent errors in the filesystem from compromising the machine
> > +	FSType string `json:"fsType,omitempty"`
> > +	// Optional: RBDPool is the rados pool name,default is rbd
> > +	RBDPool string `json:"pool"`
> > +	// Optional: RadosUser is the rados user name, default is admin
> > +	RadosUser string `json:"user"`
> > +	// Optional: Keyring is the path to key ring for RBDUser, default is /etc/ceph/keyring
> > +	Keyring string `json:"keyring"`
> > +	// Optional: Secret is the authentication secret for RBDUser, default is empty.
> > +	Secret string `json:"secret" description:"secret is authentication secret. If provided, it overrides keyring"`
> Thanks @liewegas.
> 
> @rootfs is telling me that an admin can provision a Ceph volume of, say, 5gb size, get an ID for it, and mount it via that ID in the Kubelet. So long as pod authors cannot traverse outside of that volume and read other people's data, then I think we're covered for a PV. This requires no Secret data. Just an ID for the Ceph volume that we can post as part of the PV.
> 
The kubelet has no innate permission to connect to a given ceph/gluster cluster.  If we were to require that to be specified as part of the kubelet, you would be limited to a single ceph/gluster/NFS server per cluster which is not desirable.

The alternative is that a secret has to flow with the specific PV, which means the Claim is what gives them the kubelet (on their behalf) to access the secret.  It's possible that we have to solve this the same way that pod secrets would be proposed to be secured, which is to expose the secrets for a given pod or given pvclaim as a subresource of a pod.  In that model, there is no copy of the secrets into the users namespace, because the claim links to a volume which references a secret, and so the claim authorizes the kubelet by proxy to utilize that secret.

But, for the normal non PV model, the secret still has to be accessible, so I would prefer to follow our existing secret conventions.
> When the user gets a volume via their PVClaim, they won't know or care that it's Ceph (or NFS or GCE or whatever).
> 
> Is this correct?
> 
> —
> Reply to this email directly or view it on GitHub.
>